### ISSUE TITLE
Crash in -setLastFolder (overreleased object)

### ISSUE THREAD

**Issue by stevesparks** [MailCore/mailcore2#1097]
I am making a large number of `MCOIMAPFolderStatusOperation` objects; I have 7,000 folders in my test account. Occasionally as I spin through the list I will encounter this crash.

![screen shot 2015-04-01 at 9 08 05 am](https://cloud.githubusercontent.com/assets/1730259/6941519/acc7cb4e-d84e-11e4-8f5c-dea61286e0ab.jpg)

I have looked through the issues and see that you say these operations should be run serially, and indeed, that is what I expected to be happening. But as you can see in the screenshot, this is executing on the `com.apple.root.default-qos` queue, a concurrent queue.

I noticed that the line at `MCIMAPAsyncConnection.m:96` instantiates a new operation queue for this operation. Without digging deeply into this code (I have plenty of my own fish to fry, of course) could you tell me if I am on the right trail?

Thanks,
Steve

**Comment by dinhviethoa** [comment]
If you're calling mailcore2 API in a dispatch queue, you have to set a dispatch queue on the IMAP session.
https://github.com/MailCore/mailcore2/blob/master/src/objc/imap/MCOIMAPSession.h#L122

**Comment by dinhviethoa** [comment]
Please reopen the issue if it doesn't fix your crash.

**Comment by stevesparks** [comment]
It didn't fix the crash. I added:

```objc
...
    self.imapSession = [[MCOIMAPSession alloc] init];
    _imapSession.dispatchQueue = self.imapQueue;
...

- (dispatch_queue_t)imapQueue {
    if(!_imapQueue) {
        NSString *queueName = [NSString stringWithFormat:@"IMAP queue: %@", self.emailAddress];
        _imapQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSUTF8StringEncoding], DISPATCH_QUEUE_SERIAL);
    }
    return _imapQueue;
}
```

it crashes in the same place:

![screen shot 2015-04-02 at 8 21 34 am](https://cloud.githubusercontent.com/assets/1730259/6963817/564868c8-d911-11e4-8433-a57fc606e9ea.jpg)

Note that it is still on the originating thread. Is looks like there's a race through `MC_SAFE_REPLACE_COPY`. Can we synchronize this setter?

**Comment by stevesparks** [comment]
I have no option to reopen the issue. :crying_cat_face:

**Comment by stevesparks** [comment]
This code 
```objc
    MCOIMAPFolderStatusOperation* folderInfo = [session folderStatusOperation:folderName];
    @synchronized(session) {
        [folderInfo start:^(NSError *error, MCOIMAPFolderStatus *info) {
        …
    }
```

**Comment by dinhviethoa** [comment]
Let me rephrase.
It sounds like you're starting the operation from a dispatch queue.
Is it the same queue that you're setting on the session?

The code you shown me didn't feel like you're using the same exact queue for both.

btw, the easier option for you would be to do everything from the main thread.
mailcore2 is all asycnhronous.

**Comment by stevesparks** [comment]
It is not the same queue. The UI dispatches a command to a `DISPATCH_QUEUE_PRIORITY_DEFAULT`, which then creates the Mailcore library. Are you saying that I should `dispatch_async(_session.dispatchQueue, …)` ?

**Comment by dinhviethoa** [comment]
Here's a suggestion of pattern. Does it match your codebase?

```
dispatch_queue_t s_queue;
MCOIMAPSession * s_session;

void UI_code_init()
{
   s_queue = dispatch_queue_create(...);
   dispatch_async(s_queue, ^{
      s_session = [[MCOIMAPSession alloc] init];
      s_session.dispatchQueue  = s_queue;
   });
}

void startTheOperation()
{
   dispatch_async(s_queue, ^{
     Operation * op = [s_session statusOperationWithFolder:@"INBOX"];
     [op start:^{
        // this block will be called on s_queue to ensure thread-safety.
     }];
  }
}
```

**Comment by stevesparks** [comment]
No, that's not our pattern. Obviously I have a lot of refactoring in front of me.

**Comment by dinhviethoa** [comment]
Again, the easiest way to use mailcore2 would be to call all the APIs from the main thread.
It could be done with `-performSelectorOnMainThread` (probably not realistic in regards to your actual code).

**Comment by stevesparks** [comment]
As it turns out our model is not too terrible for putting the right operations on the right queues. 

You can close this issue. It might be nice, as an enhancement, to have the assertion happen when you try to call `-start` on an inappropriate queue. But I don't expect you to undertake that.

Thanks,
Steve

**Comment by dinhviethoa** [comment]
It makes it hard :(
```
DEPRECATED FUNCTIONS
     The following functions are deprecated and will be removed in a future
     release:

     dispatch_queue_t dispatch_get_current_queue(void);
```

**Comment by stevesparks** [comment]
:( I feel your pain, brother, I feel your pain.
